File: src/server/crdt/generic/ccounter_test.go
package crdt

import (
	"reflect"
	"testing"
)

func TestCCounter_NewCCounter(t *testing.T) {
	counter := NewCCounter("node1")
	if counter == nil {
		t.Fatal("NewCCounter() returned nil")
	}
	if counter.replicaID != "node1" {
		t.Errorf("Expected counter.id to be 'node1', got %s", counter.replicaID)
	}
	if counter.Read() != 0 {
		t.Errorf("Expected initial counter value to be 0, got %d", counter.Read())
	}
}

func TestCCounter_Inc(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc2 := cc1.Clone()

	delta := cc1.Inc(5)
	t.Log(cc1)

	if cc1.Read() != 5 {
		t.Errorf("Expected counter value to be 5, got %d", cc1.Read())
	}

	cc2.Join(delta)
	if !reflect.DeepEqual(cc1, cc2) {
		t.Errorf("Expected cc1 and cc2 to be equal after joining with delta")
	}
}

func TestCCounter_IncMultiple(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc2 := cc1.Clone()
	delta1 := cc1.Inc(3)
	delta2 := cc1.Inc(7)

	if cc1.Read() != 10 {
		t.Errorf("Expected counter value to be 10 after multiple increments, got %d", cc1.Read())
	}

	cc2.Join(delta1)
	cc2.Join(delta2)
	if !reflect.DeepEqual(cc1, cc2) {
		t.Errorf("Expected cc1 and cc2 to be equal after joining with deltas")
	}
}

func TestCCounter_Dec(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc2 := cc1.Clone()

	delta := cc1.Inc(-4)

	if cc1.Read() != -4 {
		t.Errorf("Expected counter value to be -4, got %d", cc1.Read())
	}

	cc2.Join(delta)
	if !reflect.DeepEqual(cc1, cc2) {
		t.Errorf("Expected cc1 and cc2 to be equal after joining with delta")
	}
}

func TestCCounter_DecMultiple(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc2 := cc1.Clone()
	delta1 := cc1.Inc(-2)
	delta2 := cc1.Inc(-3)

	if cc1.Read() != -5 {
		t.Errorf("Expected counter value to be -5 after multiple decrements, got %d", cc1.Read())
	}

	cc2.Join(delta1)
	cc2.Join(delta2)
	if !reflect.DeepEqual(cc1, cc2) {
		t.Errorf("Expected cc1 and cc2 to be equal after joining with deltas")
	}
}

func TestCCounter_IncDec(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc2 := cc1.Clone()
	delta1 := cc1.Inc(10)
	delta2 := cc1.Inc(-4)

	if cc1.Read() != 6 {
		t.Errorf("Expected counter value to be 6 after increment and decrement, got %d", cc1.Read())
	}

	cc2.Join(delta1)
	cc2.Join(delta2)
	if !reflect.DeepEqual(cc1, cc2) {
		t.Errorf("Expected cc1 and cc2 to be equal after joining with deltas")
	}
}

func TestCCounter_IncDecToZero(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc2 := cc1.Clone()
	delta1 := cc1.Inc(7)
	delta2 := cc1.Inc(-7)

	if cc1.Read() != 0 {
		t.Errorf("Expected counter value to be 0 after incrementing and decrementing to zero, got %d", cc1.Read())
	}

	cc2.Join(delta1)
	cc2.Join(delta2)
	if !reflect.DeepEqual(cc1, cc2) {
		t.Errorf("Expected cc1 and cc2 to be equal after joining with deltas")
	}
}

func TestCCounter_IncComplex(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc2 := cc1.Clone()

	// Set to a positive value
	delta1 := cc1.Inc(10)
	if cc1.Read() != 10 {
		t.Errorf("Expected counter value to be 10 after Set(10), got %d", cc1.Read())
	}

	// Set to a lower value
	delta2 := cc1.Inc(-5)
	if cc1.Read() != 5 {
		t.Errorf("Expected counter value to be 5 after Set(5), got %d", cc1.Read())
	}

	// Set to the same value
	delta3 := cc1.Inc(0)
	if cc1.Read() != 5 {
		t.Errorf("Expected counter value to remain 5 after Set(5), got %d", cc1.Read())
	}

	// Set to zero
	delta4 := cc1.Inc(-5)
	if cc1.Read() != 0 {
		t.Errorf("Expected counter value to be 0 after Set(0), got %d", cc1.Read())
	}

	// Set to a negative value (should decrement)
	delta5 := cc1.Inc(-5)
	if cc1.Read() != -5 {
		t.Errorf("Expected counter value to be -5 after Set(-5), got %d", cc1.Read())
	}

	cc2.Join(delta1)
	t.Logf("After delta1 join, cc2: %v", cc2)
	cc2.Join(delta2)
	t.Logf("After delta2 join, cc2: %v", cc2)
	cc2.Join(delta3)
	t.Logf("After delta3 join, cc2: %v", cc2)
	cc2.Join(delta4)
	t.Logf("After delta4 join, cc2: %v", cc2)
	cc2.Join(delta5)
	t.Logf("After delta5 join, cc2: %v", cc2)

	if !reflect.DeepEqual(cc1, cc2) {
		t.Logf("cc1: %v", cc1)
		t.Logf("cc2: %v", cc2)
		t.Logf("cc2 value: %v", cc2.Read())
		t.Errorf("Expected cc1 and cc2 to be equal after joining with all deltas")
	}
}

func TestCCounter_Reset(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc1.Inc(10)
	cc2 := cc1.Clone()

	delta := cc1.Reset()

	if cc1.Read() != 0 {
		t.Errorf("Expected counter value to be 0 after reset, got %d", cc1.Read())
	}

	cc2.Join(delta)
	if !reflect.DeepEqual(cc1, cc2) {
		t.Errorf("Expected cc1 and cc2 to be equal after joining with delta")
	}
}

func TestCCounter_ResetAfterIncDec(t *testing.T) {
	counter := NewCCounter("replica1")
	counter.Inc(15)
	counter.Inc(-5)
	counter.Reset()

	if counter.Read() != 0 {
		t.Errorf("Expected counter value to be 0 after reset following increments and decrements, got %d", counter.Read())
	}
}

func TestCCounter_ResetEmpty(t *testing.T) {
	counter := NewCCounter("replica1")
	counter.Reset()

	if counter.Read() != 0 {
		t.Errorf("Expected counter value to remain 0 after resetting an empty counter, got %d", counter.Read())
	}
}

func TestCCounter_Join(t *testing.T) {
	counter1 := NewCCounter("replica1")
	counter1.Inc(5)

	counter2 := NewCCounter("replica2")
	counter2.Inc(10)

	counter1.Join(counter2)

	if counter1.Read() != 15 {
		t.Errorf("Expected counter1 value to be 15 after join, got %d", counter1.Read())
	}
	if counter2.Read() != 10 {
		t.Errorf("Expected counter2 value to remain 10, got %d", counter2.Read())
	}
}

func TestCCounter_JoinWithConflicts(t *testing.T) {
	cc1 := NewCCounter("replica1")
	cc1.Inc(5)

	cc2 := NewCCounter("replica2")
	cc2.Inc(10)

	cc1.Join(cc2)

	if cc1.Read() != 15 {
		t.Errorf("Expected counter1 value to be 15 after join, got %d", cc1.Read())
	}
	if cc2.Read() != 10 {
		t.Errorf("Expected counter2 value to remain 10, got %d", cc2.Read())
	}
}

func TestCCounter_ConcurrentUpdates(t *testing.T) {
	counter1 := NewCCounter("replica1")
	counter2 := NewCCounter("replica2")

	delta1 := counter1.Inc(5)
	delta2 := counter2.Inc(10)

	counter1.Join(delta2)
	counter2.Join(delta1)

	if counter1.Read() != 15 {
		t.Errorf("Expected counter1 value to be 15 after concurrent updates, got %d", counter1.Read())
	}
	if counter2.Read() != 15 {
		t.Errorf("Expected counter2 value to be 15 after concurrent updates, got %d", counter2.Read())
	}
}

func TestCCounter_JoinWithEmptyCounter(t *testing.T) {
	counter1 := NewCCounter("replica1")
	counter1.Inc(5)

	counter2 := NewCCounter("replica2")

	counter1.Join(counter2)

	if counter1.Read() != 5 {
		t.Errorf("Expected counter1 value to remain 5 after joining with empty counter, got %d", counter1.Read())
	}
	if counter2.Read() != 0 {
		t.Errorf("Expected counter2 value to remain 0, got %d", counter2.Read())
	}
}

File: src/client/types/crdt/generic/ccounter.test.ts
import CCounter from './ccounter';

describe('CCounter', () => {
  it('should create a new CCounter', () => {
    const counter = new CCounter('node1');
    expect(counter).not.toBeNull();
    expect(counter['replicaId']).toBe('node1');
    expect(counter.read()).toBe(0);
  });

  it('should increment the counter', () => {
    const cc1 = new CCounter('replica1');
    const cc2 = cc1.clone();

    const delta = cc1.inc(5);

    expect(cc1.read()).toBe(5);

    cc2.join(delta);
    expect(cc1).toEqual(cc2);
  });

  it('should handle multiple increments', () => {
    const cc1 = new CCounter('replica1');
    const cc2 = cc1.clone();

    const delta1 = cc1.inc(3);
    const delta2 = cc1.inc(7);

    expect(cc1.read()).toBe(10);

    cc2.join(delta1);
    cc2.join(delta2);
    expect(cc1).toEqual(cc2);
  });

  it('should decrement the counter', () => {
    const cc1 = new CCounter('replica1');
    const cc2 = cc1.clone();

    const delta = cc1.inc(-4);

    expect(cc1.read()).toBe(-4);

    cc2.join(delta);
    expect(cc1).toEqual(cc2);
  });

  it('should handle multiple decrements', () => {
    const cc1 = new CCounter('replica1');
    const cc2 = cc1.clone();

    const delta1 = cc1.inc(-2);
    const delta2 = cc1.inc(-3);

    expect(cc1.read()).toBe(-5);

    cc2.join(delta1);
    cc2.join(delta2);
    expect(cc1).toEqual(cc2);
  });

  it('should handle increment and decrement', () => {
    const cc1 = new CCounter('replica1');
    const cc2 = cc1.clone();

    const delta1 = cc1.inc(10);
    const delta2 = cc1.inc(-4);

    expect(cc1.read()).toBe(6);

    cc2.join(delta1);
    cc2.join(delta2);
    expect(cc1).toEqual(cc2);
  });

  it('should increment and decrement to zero', () => {
    const cc1 = new CCounter('replica1');
    const cc2 = cc1.clone();

    const delta1 = cc1.inc(7);
    const delta2 = cc1.inc(-7);

    expect(cc1.read()).toBe(0);

    cc2.join(delta1);
    cc2.join(delta2);
    expect(cc1).toEqual(cc2);
  });

  it('should handle complex increments and decrements', () => {
    const cc1 = new CCounter('replica1');
    const cc2 = cc1.clone();

    const delta1 = cc1.inc(10);
    expect(cc1.read()).toBe(10);

    const delta2 = cc1.inc(-5);
    expect(cc1.read()).toBe(5);

    const delta3 = cc1.inc(0);
    expect(cc1.read()).toBe(5);

    const delta4 = cc1.inc(-5);
    expect(cc1.read()).toBe(0);

    const delta5 = cc1.inc(-5);
    expect(cc1.read()).toBe(-5);

    cc2.join(delta1);
    cc2.join(delta2);
    cc2.join(delta3);
    cc2.join(delta4);
    cc2.join(delta5);

    expect(cc1).toEqual(cc2);
  });

  it('should reset the counter', () => {
    const cc1 = new CCounter('replica1');
    cc1.inc(10);
    const cc2 = cc1.clone();

    const delta = cc1.reset();

    expect(cc1.read()).toBe(0);

    cc2.join(delta);
    expect(cc1).toEqual(cc2);
  });

  it('should reset after increments and decrements', () => {
    const counter = new CCounter('replica1');
    counter.inc(15);
    counter.inc(-5);
    counter.reset();

    expect(counter.read()).toBe(0);
  });

  it('should reset an empty counter', () => {
    const counter = new CCounter('replica1');
    counter.reset();

    expect(counter.read()).toBe(0);
  });

  it('should join two counters', () => {
    const counter1 = new CCounter('replica1');
    counter1.inc(5);

    const counter2 = new CCounter('replica2');
    counter2.inc(10);

    counter1.join(counter2);

    expect(counter1.read()).toBe(15);
    expect(counter2.read()).toBe(10);
  });

  it('should join counters with conflicts', () => {
    const cc1 = new CCounter('replica1');
    cc1.inc(5);

    const cc2 = new CCounter('replica2');
    cc2.inc(10);

    cc1.join(cc2);

    expect(cc1.read()).toBe(15);
    expect(cc2.read()).toBe(10);
  });

  it('should handle concurrent updates', () => {
    const counter1 = new CCounter('replica1');
    const counter2 = new CCounter('replica2');

    const delta1 = counter1.inc(5);
    const delta2 = counter2.inc(10);

    counter1.join(delta2);
    counter2.join(delta1);

    expect(counter1.read()).toBe(15);
    expect(counter2.read()).toBe(15);
  });

  it('should join with an empty counter', () => {
    const counter1 = new CCounter('replica1');
    counter1.inc(5);

    const counter2 = new CCounter('replica2');

    counter1.join(counter2);

    expect(counter1.read()).toBe(5);
    expect(counter2.read()).toBe(0);
  });
});